fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/rdf/Skolemizer.java

```diff
@@ -0,0 +1,126 @@
+/**
+ * Copyright 2015 DuraSpace, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.fcrepo.kernel.impl.rdf;
+
+import static com.google.common.hash.Hashing.murmur3_32;
+import static com.hp.hpl.jena.rdf.model.ResourceFactory.createResource;
+import static com.hp.hpl.jena.rdf.model.ResourceFactory.createStatement;
+import static java.util.UUID.randomUUID;
+import static org.slf4j.LoggerFactory.getLogger;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import org.slf4j.Logger;
+
+import com.google.common.base.Function;
+import com.google.common.base.Supplier;
+import com.google.common.hash.HashCode;
+import com.hp.hpl.jena.rdf.model.AnonId;
+import com.hp.hpl.jena.rdf.model.Literal;
+import com.hp.hpl.jena.rdf.model.Property;
+import com.hp.hpl.jena.rdf.model.RDFNode;
+import com.hp.hpl.jena.rdf.model.RDFVisitor;
+import com.hp.hpl.jena.rdf.model.Resource;
+import com.hp.hpl.jena.rdf.model.Statement;
+
+/**
+ * Skolemization is abstractly a function from RDF nodes to RDF nodes, but here we implement it, purely for
+ * convenience of operation, as a function from triples to triples. {@link #nodeSkolemizer} represents the
+ * node-to-node function. An instance of this class will make two blank nodes with the same label into the same Skolem
+ * resource because it uses only one prefix for its lifetime and because the mapping from a given blank node to a
+ * Skolem resource depends only on that prefix and on a MurmurHash3 of the label of the blank node in hand. An
+ * instance of this class should be used only with a contextual topic of one single resource and only for one
+ * document's scope of RDF about that resource.
+ *
+ * @author ajs6f
+ */
+public class Skolemizer implements Function<Statement, Statement>, Supplier<Set<Resource>> {
+
+    private final RDFVisitor nodeSkolemizer;
+
+    private final Set<Resource> skolemNodes = new HashSet<>();
+
+    private static final Logger log = getLogger(Skolemizer.class);
+
+    /**
+     * @param topic The URI of the contextual topic of the RDF supplied to this code. Blank nodes will be skolemized
+     *        to identifiers rooted at this URI.
+     */
+    public Skolemizer(final Resource topic) {
+        final String prefix = topic + "/" + randomUUID().toString().replace('-', '/');
```

Comment:
I have not probed more deeply yet, but I am concerned that `Resource.toString()` here will create a String that is not compatible with JCR path creation performed in `JcrPropertyStatementListener.addedStatement()`.
https://github.com/apache/jena/blob/master/jena-core/src/main/java/com/hp/hpl/jena/rdf/model/Resource.java#L95-L104

Hopefully this is a non-issue.